### PR TITLE
Blog contribution: article about type-safe error handling using typescript, codegen and fp-ts

### PR DESCRIPTION
**Pull request by gthau**
Hi, 

following @dotansimha suggestion, I'm proposing this blog post.
This is an article that explains the approach showcased in this repository: https://github.com/gthau/graphql-error-handling-with-union-and-fpts 

Maybe the first part of the article could be summarized since it's basically reexplaining the schema driven error handling already described in Sasha Solomon's medium: [200 OK! Error Handling in GraphQL](https://sachee.medium.com/200-ok-error-handling-in-graphql-7ec869aec9bc) and @n1ru4l's [Handling GraphQL errors like a champ with unions and interfaces](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/).

The rest of the article describes how to take this approach one step further with typescript and codegen. And another step further with fp-ts.

Any suggestion is welcome :)

### CODE REVIEW

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+- 401 or 403: unauthorized (need authentication) or forbidden (insufficient permission)
+- 404: resource not found
+- 500: internal server error
+- and more.
+
+
+# GraphQL default error handling
+
+
+In GraphQL, you'll get a 5xx error in case the server is not available at all, but for other exceptions you'll get the standard error handling mechanism from GraphQL: in the JSON you receive there is:
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:25]:
```suggestion
In GraphQL, all requests must go to a single endpoint:


> POST https://example.com/graphql

A GraphQL API will always return a `200 OK` Status Code, even in case of error.
You'll get a 5xx error in case the server is not available at all.

The standard error handling mechanism from GraphQL with return a JSON containing:
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+- a `data` key which contains the data corresponding to the GraphQL operation (query, mutation, subscription) invoked and
+- an `errors` key which contains an array of errors returned by the server, with a message and location.
+
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:28]: Maybe add the following below?

"
Example of error result:

```json
{
  "errors": [
    {
      "message": "User not found",
      "locations": [{ "line": 6, "column": 7 }],
      "path": ["user", 1]
    }
  ]
}
```
"

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+This solution is however not satisfactory for the following reasons:
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:29]:
```suggestion
This standard error handling is not satisfactory for the many reasons:
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+1. it is not easily parseable, it is more meant for a human user to read
+2. it is not typed: you can add more information, e.g. an error code and other fields, but those are not part of the schema
+3. it is not self-documented: by looking at the operation signature, you don't know what might fail
+4. it treats some domain results as errors / exceptions whereas they are just alternate results of the operation. E.g. an entity not found for a given id and a user not having permission to access an entity belong to the domain model, whereas runtime exceptions such as the GraphQL server failing to communicate with the database / external microservice, or an operation timing out because of slowness are real exceptions: we don't get any result that relates to domain model, just an unexpected runtime error.
+
+
+# Better modelisation of errors using Union and Interfaces
+
+
+GraphQL has support for Union types (only for Types, not yet for Inputs). We can take advantage of it and design our schema to specify the happy and unhappy paths.
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:40]: What do "happy and unhappy paths" mean?
Is there a clearer expression we could use? (ex: "expected and erroneous results")

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:41]: Here again, a query + response example would make it easier the benefits listed below:

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+Benefits:
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:42]:
```suggestion
Leveraging GraphQL Union types for error cases brings many benefits:
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+
+1. Type-safety: the errors are also typed
+2. The consumer cannot ignore the errors
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:45]:
```suggestion
2. The consumer cannot ignore the errors (the must be handled using inline fragments)
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+3. Self-documented: the operation signature includes all possible cases (result and errors) therefore less documentation is required to explain the possible error cases
+4. The _unhappy_ paths are not unexpected anymore, there are just other possible results
+
+## Schema modification
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:49]:
```suggestion
## Implementing Union types based error handling
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+
+Given a schema like the following:
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:51]:
```suggestion
Given the following Schema:
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+type Query {
+  entity(id: ID!, userId: ID!): Entity
+}
+```
+
+If something fails during the execution, we can:
+- throw an Error and use the default GraphQL error mechanism explained above
+- return `null`: this is not satisfactory because we can't know the reason the operation returned `null`
+
+Let's make the following modification to handle all possible paths:
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:68]:
```suggestion
Let's instead implement error handling using Union types with the following modifications:
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+  | NotAllowedError
+  | InvalidInputError
+  | UnknownError
+
+type Query {
+  entity(id: ID!, userId: ID!): EntityResult!
+}
+```
+
+We define a `BaseError` interface that all concrete Errors implement. And we define the query result as a Union type of the happy path (an `Entity`) and the other possible results (`NotFoundError`, `NotAllowedError`, etc.). In this manner we extensively describe all possible results, with the added benefit that we can make our result type non-nullable (with the `!` character).
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:108]:
```suggestion
We define a `BaseError` interface that all concrete Errors implement.
Then, we define the query result as a Union type of the happy path (an `Entity`) and the other possible results (`NotFoundError`, `NotAllowedError`, etc.). In this manner, we extensively describe all possible results, with the added benefit that we can make our result type non-nullable (with the `!` character).
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+  },
+};
+```
+
+Since we include the `__typename` and the fields match the Schema type definitions, there is no need to define type resolvers.
+
+For more details on this solution, you can read this article from [Laurin Quast](https://github.com/n1ru4l): [Handling GraphQL errors like a champ with unions and interfaces](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/).
+
+
+# Typesafe error handling
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:151]:
```suggestion
# Typesafe error handling on the client-side
```

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 13]
```diff
@@ -0,0 +1,631 @@
...
+
+
+Let's now take it to the next level using Typescript and GraphQL Code Generator.
+
+[Typescript](https://www.typescriptlang.org/) is a superset of Javascript which adds type annotations to Javascript code and therefore type safety (with caveats) at compile-time.
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:156]: We can assume that the audience knows TypeScript and remove this paragraph.

(Last State of JS survey shows that most people know it or are using it: https://2021.stateofjs.com/en-US/other-tools/#javascript_flavors

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 14]
```diff
@@ -0,0 +1,631 @@
...
+    plugins:
+      - add:
+          content: '/* eslint-disable */'
+      - add:
+          content: "import * as E from './src/errors';"
+      - "typescript"
+      - "typescript-resolvers"
+    config:
+      mappers:
+        InputFieldValidation: 'E.InputFieldError'
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:209]: nice! ⚡

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+
+:::note
+We use the `add` plugin to import the `WrappedError` generic wrapper type and the Errors types. If we don't do that, we would have to define the full path for each mapper, e.g.:
+
+```yaml
+NotFoundError: 'import("./src/errors").WrappedError<import("./src/errors").NotFoundError>'
+```
+:::
+
+This `codegen` configuration generates the Typescript types from the Schema and the proper signatures for the operations (query/mutation/subscription) and types resolvers. Now we have type safety at compile-time and type inference and hints in the code editor.
```
Comment (charlypoly) [inline pages/blog/graphql-error-handling-with-fp.mdx:225]:
```suggestion
This `codegen` configuration generates the Typescript types from the Schema and the proper signatures for the operations (query/mutation/subscription) and types resolvers. Now we have static type safety and type inference and hints in the code editor.
```

**Review by charlypoly — COMMENTED** [review]
@gthau Thank you for your article, that's great!

I think we can leave the introduction as it is since it is pretty important to understand the subject of the whole article.

I've left some suggestions 👀

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 4 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:40]: Happy path = the expected result. Designing for the happy path is living in a world where nothing unexpected will happen, and obviously will lead to runtime errors.
The "unhappy path" are the alternative paths that exist in addition to the expected result (happy path).
It's quite common to come across this term, e.g. https://en.wikipedia.org/wiki/Happy_path#Unhappy_path, https://cucumber.io/blog/test-automation/happy-unhappy-paths-why-you-need-to-test-both/

I can reword it. We could say "the expected and alternative results"

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 5 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:49]: Sure, will add one

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 4 above)
Comment (charlypoly) [reply pages/blog/graphql-error-handling-with-fp.mdx:48]: Okay, I see!

I feel like it's a vocabulary mainly from the TDD/BDD world that most people won't get properly (a bit too theoretical)

Rewording it would make it easier to follow for a broader audience 🤓

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 16]
```diff
@@ -0,0 +1,640 @@
...
+- 404: resource not found
+- 500: internal server error
+- and more.
+
+
+# GraphQL default error handling
+
+
+In GraphQL, all requests must go to a single endpoint:
+
```
Comment (Urigo) [inline pages/blog/graphql-error-handling-with-fp.mdx:26]: You could still actually do that no?
Send 400 and other status codes, based on queries.
I'm not saying it's a good practice but it is possible and that also helps when using old school tools for monitoring that don't understand GraphQL

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 17]
```diff
@@ -0,0 +1,640 @@
...
+1. it is not easily parseable, it is more meant for a human user to read
+2. it is not typed: you can add more information, e.g. an error code and other fields, but those are not part of the schema
+3. it is not self-documented: by looking at the operation signature, you don't know what might fail
+4. it treats some domain results as errors / exceptions whereas they are just alternate results of the operation. E.g. an entity not found for a given id and a user not having permission to access an entity belong to the domain model, whereas runtime exceptions such as the GraphQL server failing to communicate with the database / external microservice, or an operation timing out because of slowness are real exceptions: we don't get any result that relates to domain model, just an unexpected runtime error.
+
+
+# Better modelisation of errors using Union and Interfaces
+
+
+GraphQL has support for Union types (only for Types, not yet for Inputs). We can take advantage of it and design our schema to specify the happy and unhappy paths.
```
Comment (Urigo) [inline pages/blog/graphql-error-handling-with-fp.mdx]: maybe worth linking into that - https://github.com/graphql/graphql-spec/issues/488#issuecomment-982901990

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 18]
```diff
@@ -0,0 +1,640 @@
...
+      };
+    },
+  },
+};
+```
+
+Since we include the `__typename` and the fields match the Schema type definitions, there is no need to define type resolvers.
+
+For more details on this solution, you can read this article from [Laurin Quast](https://github.com/n1ru4l): [Handling GraphQL errors like a champ with unions and interfaces](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/).
+
```
Comment (Urigo) [inline pages/blog/graphql-error-handling-with-fp.mdx:226]: Maybe worth adding here also the possible OneOf solution:
https://github.com/graphql/graphql-spec/pull/825

and how to use it already today: https://github.com/graphql/graphql-spec/pull/825#issuecomment-1039997787

Comment (Urigo) [reply pages/blog/graphql-error-handling-with-fp.mdx:226]: ## OneOf

If your GraphQL setup supports `@oneOf` (like with Envelop), you can use it for that.

oneOf could completely replace this and makes `__typename` obsolete.

You basically have to decide: do you want an error *or* the payload, or do you want the payload *and* errors. I tend to do the latter, the existence of a payload isn't sufficient to indicate success - you need to request something (the thing you created, a success boolean, whatever) to see if it succeeded. This is the best way, IMO, of giving the user type-safe details of what went wrong whilst still allowing for partial success.

e.g. "saveAndPublish" might successfully save but fail to publish; we might not want to roll both back

Another thing that might be interesting for this is literal enums.

```graphql
enum UserResultType {
  error
  success
}

type UserSuccess {
  type: UserResultType.success
}

type UserError {
  type: UserResultType.error
}

union UserResult = UserSuccess | UserError
```

or you could use an interface

```graphql
type UserSuccess implements UserResult {
  type: UserResultType.success
}

type UserError implements UserResult  {
  type: UserResultType.error
}

interface UserResult {
  type: UserResultType
}
```

and no `__typename` needed and Codegen could then benefit from this.

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 19]
```diff
@@ -0,0 +1,640 @@
...
+
+# Typesafe error handling on the client-side
+
+
+Let's now take it to the next level using Typescript and GraphQL Code Generator.
+
+[Typescript](https://www.typescriptlang.org/) is a superset of Javascript which adds type annotations to Javascript code and therefore type safety (with caveats) at compile-time.
+
+[GraphQL Code Generator](https://www.graphql-code-generator.com/) is a tool which generates type definitions which corresponds to the GraphQL schema. It has several plugins and we'll use 3 of those:
```
Comment (Urigo) [inline pages/blog/graphql-error-handling-with-fp.mdx:233]: I would add here "developed by [The Guild](https://the-guild.dev/)"

**Review by Urigo — CHANGES_REQUESTED** [review]

**Comment by Urigo** [comment]
What a great post, thanks @gthau !

I would for this post to be the thing we would link to whenever people are asking us about error handling in GraphQL.
We have a Notion doc where we keep track of our thoughts around that, do you think you can go through that and see if you can/should cover everything in there also here and we would deprecate that doc in favor of your post?
https://theguildoss.notion.site/Error-handling-in-GraphQL-343aba136d3f4228a7d798b88bde2165

Also on the subject of creating types for the results of remote APIs/DBs, what are your thought about generating these types from schemas and things like that?
Is that worth mentioning here?

Thank you for creating this!

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 18 above)
Comment (Urigo) [reply pages/blog/graphql-error-handling-with-fp.mdx:226]: if you don't mind checking it out and if after that you would think it's valuable, I would love it if you do

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 16 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:26]: Does it make sense to mention it in the context of this article?

Do you want me to reword it to make it clearer that it's customary but not mandatory to send 200 status codes for all queries returns?

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 18 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:226]: I'll experiment with it on my repo.

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 16 above)
Comment (Urigo) [reply pages/blog/graphql-error-handling-with-fp.mdx:26]: yes, I think it's important because many people are using this argument to bash on GraphQL compared to REST, saying GraphQL can't have normal status codes that work with popular monitoring tools.
But that's false, because you can treat GraphQL like REST and have all of that, but, you can also make it much better, with the approach you are suggesting.
So it's not a downside of GraphQL, it's something that GraphQL can do but can also do much better :)

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 19 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx]: > What a great post, thanks @gthau !

And thank you for your suggestions ;)

> I would for this post to be the thing we would link to whenever people are asking us about error handling in GraphQL. We have a Notion doc where we keep track of our thoughts around that, do you think you can go through that and see if you can/should cover everything in there also here and we would deprecate that doc in favor of your post? https://theguildoss.notion.site/Error-handling-in-GraphQL-343aba136d3f4228a7d798b88bde2165

I'll go through it and check what can be incorporated.
 
> Also on the subject of creating types for the results of remote APIs/DBs, what are your thought about generating these types from schemas and things like that? Is that worth mentioning here?

Indeed usually the types will be auto-generated from Graphql schema, JSON schema for REST API or from database schema. For the purpose of the article, I wanted to concentrate on the typed error handling and the safety and API self-documentation the monads bring, so I tried to make the repo the bare minimum to showcase this.

Not sure if I should use a real API in the showcase repo to demo the remote APIs type generation. I don't know if it risks diluting the intent of the article.

Comment (Urigo) [reply pages/blog/graphql-error-handling-with-fp.mdx]: yes I agree about the type generation. maybe worth mentioning that option in a single sentence? with links to further reading?

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 18 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:226]: > If your GraphQL setup supports `@oneOf` (like with Envelop), you can use it for that.
> 
> oneOf could completely replace this and makes `__typename` obsolete.
I've tried the @oneOf directive. It only works on Input type and on Query field (in order to select one of the arguments of the query.

As it is implemented currently, I don't see a way to use it for output Types in order to substitute the use of Union.

Am I missing something?

Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:226]: > Another thing that might be interesting for this is literal enums.
> 
> ```graphql
> enum UserResultType {
>   error
>   success
> }
> 
> type UserSuccess {
>   type: UserResultType.success
> }
> 
> type UserError {
>   type: UserResultType.error
> }
> 
> union UserResult = UserSuccess | UserError
> ```
> 
> or you could use an interface
> 
> ```graphql
> type UserSuccess implements UserResult {
>   type: UserResultType.success
> }
> 
> type UserError implements UserResult  {
>   type: UserResultType.error
> }
> 
> interface UserResult {
>   type: UserResultType
> }
> ```
> and no __typename needed and Codegen could then benefit from this.

With this solution, the goal would be to improve the consumption of the query, not the server-side code, right?
e.g. instead of having typeguard client-side like:
```ts
function isGraphqlError(v: unknown): v is BaseError {
  return (v as any)?.__typename.endsWith('Error');
}
```
I would have:
```ts
function isGraphqlError(v: unknown): v is BaseError {
  return (v as any)?.type === ResultType.error;
}
```

Comment (Urigo) [reply pages/blog/graphql-error-handling-with-fp.mdx:226]: cc @n1ru4l could you validate?

File: pages/blog/graphql-error-handling-with-fp.mdx [hunk 20]
```diff
@@ -0,0 +1,706 @@
...
+This standard error handling is not satisfactory for the many reasons:
+
+1. it is not easily parseable, it is more meant for a human user to read
+2. it is not typed: you can add more information, e.g. an error code and other fields, but those are not part of the schema
+3. it is not self-documented: by looking at the operation signature, you don't know what might fail
+4. it treats some domain results as errors / exceptions whereas they are just alternate results of the operation. E.g. an entity not found for a given id and a user not having permission to access an entity belong to the domain model, whereas runtime exceptions such as the GraphQL server failing to communicate with the database / external microservice, or an operation timing out because of slowness are real exceptions: we don't get any result that relates to domain model, just an unexpected runtime error.
+
+
+# Better modelisation of errors using Union and Interfaces
+
```
Comment (Urigo) [inline pages/blog/graphql-error-handling-with-fp.mdx:58]: Let's also embed @notrab 's video in here!
https://youtu.be/AJOh-0Gdvt8

**Review by Urigo — CHANGES_REQUESTED** [review]

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 19 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx]: > yes I agree about the type generation. maybe worth mentioning that option in a single sentence? with links to further reading?

Ok, will do.

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 20 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:58]: Added at the end of the section, see commit https://github.com/the-guild-org/the-guild-website/pull/611/commits/1a5222c90c1359bd80d42193537f7936f734972a

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 19 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx]: > Also on the subject of creating types for the results of remote APIs/DBs, what are your thought about generating these types from schemas and things like that?
> Is that worth mentioning here?

Added a note about it, see commit https://github.com/the-guild-org/the-guild-website/pull/611/commits/a2787c37f6de45cea720cf0865259520583c3e77

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 16 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:26]: Looking at [graphql-helix implementation](https://github.com/contra/graphql-helix/blob/a796f9d2afdd8c2d91b745edbc53c5d49b716ba6/packages/core/lib/process-request.ts#L64), it seems that this must be handled out of the query resolvers code, at the server level, right?  One must throw an error from the query resolver (or not handle a bubbled up error), then at the server level catch and type check the possible errors to assign the "correct" http status code.

Another question: how is this managed client-side, e.g. with Apollo client? All non-200 response returns will be considered as errors? And in observable pipelines would need to be handled with an RXJS [catchError](https://rxjs.dev/api/operators/catchError) operator?

**Review by dotansimha — APPROVED** [review]

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 16 above)
Comment (Urigo) [reply pages/blog/graphql-error-handling-with-fp.mdx:26]: Yes that's true, but also maybe in other implementation it is handled differently.
I wouldn't go too deep into it because there are many different ways to implement it, I would just mention it is possible.
About the client, also depends on the client you are using, so also here I would stay high level and general

**Review by Urigo — APPROVED** [review]

File: pages/blog/graphql-error-handling-with-fp.mdx (hunk 16 above)
Comment (gthau) [reply pages/blog/graphql-error-handling-with-fp.mdx:26]: Ok, I'll phrase to mention that it's possible. And link to example of how graphql-helix does it?

Comment (Urigo) [reply pages/blog/graphql-error-handling-with-fp.mdx:26]: yes that's ok, we can always update it later with Yoga :)

**Comment by Urigo** [comment]
this is great! I think we can merge and publish once you'll say it's all ready @gthau !

**Comment by gthau** [comment]
@Urigo  I added:
- a few lines to the intro and a link to the repository
- a note about status codes (with link to graphql-helix impl)
- a note about `oneOf` and the Envelop/GraphQL Yoga plugin

Once you approved, it's good to publish

**Comment by Urigo** [comment]
@gthau looks great to me!
Merging and tomorrow we'll post on social media.
Can you send here links to your profiles on social media sites and also maybe what you wish we'll write in the posts?

**Comment by gthau** [comment]
> @gthau looks great to me! Merging and tomorrow we'll post on social media. Can you send here links to your profiles on social media sites

Social media sites: [github](https://github.com/gthau), [twitter](https://twitter.com/ghislainthau), [LinkedIn](https://www.linkedin.com/in/ghislain-thau-70860416/)

> and also maybe what you wish we'll write in the posts?

Do you have an example of how you usually announce the posts?

**Comment by dotansimha** [comment]
@Urigo build should pass now. We still missing a cover image for this blog post.

File: lib/meta.ts [hunk 21]
```diff
@@ -29,6 +29,7 @@ export function hasManyAuthors(meta: object): meta is MetaWithAuthors {
 }
 
 export function pickAuthor(meta: Meta): string {
+  console.log(meta);
```
Comment (n1ru4l) [inline lib/meta.ts:32]: 👀

Comment (Urigo) [reply lib/meta.ts:32]: I'm merging and will delete after

**Comment by Urigo** [comment]
@gthau I forgot we need a background image!

**Comment by Urigo** [comment]
here is an example tweet: https://twitter.com/whereischarly/status/1493249361194725385?s=20&t=Qoz7GFVLOm9F4QxIIjKUBQ

**Comment by gthau** [comment]
> here is an example tweet: https://twitter.com/whereischarly/status/1493249361194725385?s=20&t=Qoz7GFVLOm9F4QxIIjKUBQ

What about: _GraphQL error handling can be challenging. Checkout this article to know how to do it in a type-safe manner by leveraging the GraphQL's Type system, its rich ecosystem of tools and libraries and optionally functional programming techniques._

> @gthau I forgot we need a background image!

Illustration is not my strong point :)
What about those for the square preview and the 2nd one for the article itself?
article-small
article-large

**Comment by Urigo** [comment]
I've sent a message to our designer, hopefully tonight or tomorrow we'll have something and then post on social media tomorrow

**Comment by Urigo** [comment]
@gthau do you think this is interesting?
Do you think it's something we should add to the post?
https://twitter.com/michlbrmly/status/1501200008091246601
